pkg/reconciler/identity/iam_policy.go

```diff
@@ -0,0 +1,238 @@
+}
+
+// IAMPolicyManager serializes and batches IAM policy changes to a Google
+// Service Account to avoid conflicting changes.
+type IAMPolicyManager struct {
+	iam         gclient.IamClient
+	requestCh   chan *modificationRequest
+	pending     map[GServiceAccount]*batchedModifications
+	getPolicyCh chan *getPolicyResponse
+}
+
+// NewIAMPolicyManager creates an IAMPolicyManager using the given
+// IamClient. The IAMPolicyManager will execute until ctx is cancelled.
+func NewIAMPolicyManager(ctx context.Context, client gclient.IamClient) (*IAMPolicyManager, error) {
+	m := &IAMPolicyManager{
+		iam:         client,
+		requestCh:   make(chan *modificationRequest),
+		pending:     make(map[GServiceAccount]*batchedModifications),
+		getPolicyCh: make(chan *getPolicyResponse),
+	}
+	go m.manage(ctx)
+	return m, nil
+}
+
+// AddIAMPolicyBinding adds or updates an IAM policy binding for the given account
+// and role to include member. This call will block until the IAM update
+// succeeds or fails or until ctx is cancelled.
+func (m *IAMPolicyManager) AddIAMPolicyBinding(ctx context.Context, account GServiceAccount, member string, role iam.RoleName) error {
+	return m.doRequest(ctx, &modificationRequest{
+		serviceAccount: account,
+		role:           role,
+		member:         member,
+		action:         Add,
+		respCh:         make(chan error, 1),
+	})
+}
+
+// RemoveIAMPolicyBinding removes or updates an IAM policy binding for the given
+// account and role to remove member. This call will block until the IAM update
+// succeeds or fails or until ctx is cancelled.
+func (m *IAMPolicyManager) RemoveIAMPolicyBinding(ctx context.Context, account GServiceAccount, member string, role iam.RoleName) error {
+	return m.doRequest(ctx, &modificationRequest{
+		serviceAccount: account,
+		role:           role,
+		member:         member,
+		action:         Remove,
+		respCh:         make(chan error, 1),
+	})
+}
+
+func (m *IAMPolicyManager) doRequest(ctx context.Context, req *modificationRequest) error {
+	select {
+	case m.requestCh <- req:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	select {
+	case err := <-req.respCh:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (m *IAMPolicyManager) manage(ctx context.Context) {
+	for {
+		select {
+		case req := <-m.requestCh:
+			if err := m.makeModificationRequest(ctx, req); err != nil {
+				req.respCh <- err
+			}
+		case getPolicy := <-m.getPolicyCh:
+			batched := m.pending[getPolicy.account]
+			if len(batched.listeners) == 0 {
+				delete(m.pending, getPolicy.account)
+				break
+			}
+			if getPolicy.err != nil {
+				for _, listener := range batched.listeners {
+					listener <- getPolicy.err
+				}
+				delete(m.pending, getPolicy.account)
+				break
+			}
+			m.pending[getPolicy.account] = &batchedModifications{
+				roleModifications: make(map[iam.RoleName]*roleModification),
+			}
+			go m.applyBatchedModifications(ctx, getPolicy.account, getPolicy.policy, batched)
```

Comment:
`batched` is read in this new go routine. Do we need any concurrency control to makes sure `manage` and `makeModificationRequest` don't modify `batched` at the same time as this go routine is reading it?

Reply:
No, batched is replaced with a new empty batch in m.pending immediately prior to this so ownership of batched is effectively handed off to the applyBatchedModifications goroutine.